File: server/src/models/implement_models/general_case_models/body_composition_model.py
import datetime
from mongoengine import *
from src.constants.enums import Activity

class BodyComposition(Document):
    recordDate = DateTimeField(required=True, default=datetime.datetime.utcnow)
    height = DecimalField(required=True, precision=2, min_value=0)
    weight = DecimalField(required=True, precision=2, min_value=0)
    percentBodyFat = IntField(min_value=0, max_value=100)
    activity = EnumField(Activity, required=True)
    BMI = DecimalField(required=True, precision=2, min_value=0)
    BMR = DecimalField(required=True, precision=2, min_value=0)
    TDEE = DecimalField(required=True, precision=2, min_value=0)

    meta = {
        "db_alias": "miao",
        "collection": "body_compositions"
    }

File: server/src/models/embedded_models/bodyComposition.model.js
import mongoose from "mongoose";
import { PracticeFrequency } from "../../constants/enums.js";

const schema = new mongoose.Schema({
  recordDate: {
    type: Date,
    required: true,
    default: new Date(),
  },
  weight: {
    type: Number,
    required: true,
    default: 0,
    set: (w) => Math.round(w * 10) / 10,
    get: (w) => Math.round(w * 10) / 10,
  },
  height: {
    type: Number,
    required: true,
    default: 0,
    set: (h) => Math.round(h),
    get: (h) => Math.round(h),
  },
  percentBodyFat: {
    type: Number,
    default: null,
    min: 0,
    max: 100,
    set: (f) => Math.round(f),
    get: (f) => Math.round(f),
  },
  activity: {
    type: String,
    required: true,
    enum: [
      PracticeFrequency.RARELY,
      PracticeFrequency.OCCASIONALLY,
      PracticeFrequency.SOMETIMES,
      PracticeFrequency.NORMALLY,
      PracticeFrequency.ALWAYS,
    ],
    default: PracticeFrequency.RARELY,
  },
  BMI: {
    type: Number,
    set: (bmi) => Math.round(bmi * 10) / 10,
    get: (bmi) => Math.round(bmi * 10) / 10,
    required: true,
  },
  BMR: {
    type: Number,
    set: (bmr) => Math.round(bmr),
    get: (bmr) => Math.round(bmr),
    required: true,
  },
  TDEE: {
    type: Number,
    set: (tdee) => Math.round(tdee),
    get: (tdee) => Math.round(tdee),
    required: true,
  },
});

export default schema;
